### PR TITLE
Skip normalization & obfuscation and coalesce instead

### PR DESCRIPTION
**Pull request by bwoebi**
Turns out the agent ignores the root_span_tags for v0.7 anyway. This is useful to coalesce chunks with common data, allowing us to actually save us from doing redundant requests. As it was currently, the agentful trace sender always sent one trace per request, instead of merging them.

We also skip normalization and rely on the agent to do it, for sake of consistency with other tracers and reduce the potentially duplicated work across tracer and agent. We may change that back in future, but for now we've determined it to be the easiest way to work, also with respect to normalization-unaware testing.

### CODE REVIEW

File: trace-utils/src/trace_utils.rs [hunk 1]
```diff
@@ -103,6 +116,20 @@ pub fn coalesce_send_data(mut data: Vec<SendData>) -> Vec<SendData> {
...
     });
+    // Merge chunks with common properties. Reduces requests for agentful mode.
+    // And reduces a little bit of data for agentless.
+    for send_data in data.iter_mut() {
+        send_data
+            .tracer_payloads
+            .sort_unstable_by(cmp_send_data_payloads);
+        send_data.tracer_payloads.dedup_by(|a, b| {
+            if cmp_send_data_payloads(a, b) == Ordering::Equal {
+                b.chunks.append(&mut a.chunks);
```
Comment (pierotibou) [inline trace-utils/src/trace_utils.rs:127]:
```suggestion
        send_data.tracer_payloads.dedup_by(|a, b| {
            if cmp_send_data_payloads(a, b) == Ordering::Equal {
                // Note: dedup_by drops a, and retains b.
                b.chunks.append(&mut a.chunks);
```

File: trace-utils/src/trace_utils.rs [hunk 2]
```diff
@@ -103,6 +116,20 @@ pub fn coalesce_send_data(mut data: Vec<SendData>) -> Vec<SendData> {
         }
         false
     });
+    // Merge chunks with common properties. Reduces requests for agentful mode.
```
Comment (pierotibou) [inline trace-utils/src/trace_utils.rs:119]: Is it an optimization that we should keep for another PR? To stay on the safe side?

File: trace-utils/src/trace_utils.rs [hunk 3]
```diff
@@ -288,10 +315,12 @@ pub fn collect_trace_chunks(
     mut traces: Vec<Vec<pb::Span>>,
     tracer_header_tags: &TracerHeaderTags,
     process_chunk: impl Fn(&mut TraceChunk, usize),
+    is_agentless: bool,
 ) -> pb::TracerPayload {
     let mut trace_chunks: Vec<pb::TraceChunk> = Vec::new();
 
-    let mut gathered_root_span_tags = false;
+    // We'll skip setting the global metadata and reply on the agent to unpack these
```
Comment (pierotibou) [inline trace-utils/src/trace_utils.rs:322]:
```suggestion
    // We'll skip setting the global metadata and rely on the agent to unpack these
```

File: trace-mini-agent/src/trace_processor.rs [hunk 4]
```diff
@@ -85,6 +85,7 @@ impl TraceProcessor for ServerlessTraceProcessor {
                     obfuscate_span(span, &config.obfuscation_config);
                 }
             },
+            true,
```
Comment (pierotibou) [inline trace-mini-agent/src/trace_processor.rs:88]:
```suggestion
            true, // In mini agent, we always send agentless
```

**Review by pierotibou — COMMENTED** [review]
nitpicks on comments

File: trace-utils/src/trace_utils.rs (hunk 2 above)
Comment (bwoebi) [reply trace-utils/src/trace_utils.rs:119]: Sort of mandatory fix, yes.

**Review by pierotibou — APPROVED** [review]

**Review by hoolioh — APPROVED** [review]

File: trace-utils/Cargo.toml [hunk 5]
```diff
@@ -23,12 +23,12 @@ datadog-trace-normalization = { path = "../trace-normalization" }
 tokio = { version = "1", features = ["macros", "rt-multi-thread"] }
 rand = "0.8.5"
 bytes = "1.6.0"
-# This should only be used for testing. It isn't under dev-dependencies because test-utils can't be under #[cfg(test)].
```
Comment (ekump) [inline trace-utils/Cargo.toml:26]: This needs to be put back.

**Comment by ekump** [comment]
bypassing branch protection and merging at @pierotibou's request. dd-gitlab/report_gitlab_CI_status job is not working as expected. Gitlab manually verified to pass.
